# 予測関数の実装
MNISTの学習済み重みを用いて、予測(クラス分類を行うこと)関数を実装していく

In [1]:
try:
    from google.colab import files
    print('Google Colab. 上での実行です')
    print('「ファイルを選択」から、DAY1/2_notebookフォルダのsample_weight.pklを選択し、アップロードしてください')
    print('===========')
    files.upload()
except:
    print('ローカル環境での実行です')

ローカル環境での実行です


In [2]:
%matplotlib inline
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
import pickle
from sklearn.metrics import accuracy_score

# Load the MNIST dataset
import tensorflow as tf
mnist = tf.keras.datasets.mnist
(X_train, y_train),(X_test, y_test) = mnist.load_data()

from sklearn.preprocessing import LabelBinarizer
lb = LabelBinarizer()

train = X_train/255
test = X_test/255
train = train.reshape(-1, 28*28)
test = test.reshape(-1, 28*28)
train_labels = lb.fit_transform(y_train)
test_labels = lb.fit_transform(y_test)

2022-09-07 03:35:27.199470: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-07 03:35:27.444235: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-09-07 03:35:27.444273: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2022-09-07 03:35:27.478796: E tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2022-09-07 03:35:28.274687: W tensorflow/stream_executor/platform/de

11490434/11490434 [==============================] - 0s 0us/step


## 学習済みの重みを読む

In [3]:
def init_network():
    with open("sample_weight.pkl", "rb") as f:
        network = pickle.load(f)
    return network

network = init_network()
print(network.keys())
print("")

W1, W2, W3 = network["W1"],network["W2"],network["W3"]
b1, b2, b3 = network["b1"],network["b2"],network["b3"] 

print(network["W1"])

dict_keys(['b2', 'W1', 'b1', 'W2', 'W3', 'b3'])

[[-0.00741249 -0.00790439 -0.01307499 ...  0.01978721 -0.04331266
  -0.01350104]
 [-0.01029745 -0.01616653 -0.01228376 ...  0.01920228  0.02809811
   0.01450908]
 [-0.01309184 -0.00244747 -0.0177224  ...  0.00944778  0.01387301
   0.03393568]
 ...
 [ 0.02242565 -0.0296145  -0.06326169 ... -0.01012643  0.01120969
   0.01027199]
 [-0.00761533  0.02028973 -0.01498873 ...  0.02735376 -0.01229855
   0.02407041]
 [ 0.00027915 -0.06848375  0.00911191 ... -0.03183098  0.00743086
  -0.04021148]]


### 行列の形状の確認

In [4]:
print(W1.shape, W2.shape, W3.shape)
print("            " , b1.shape, "      " ,b2.shape,"    " ,  b3.shape)

(784, 50) (50, 100) (100, 10)
             (50,)        (100,)      (10,)


### [問]
* 入力層のノード数は？
    * 784個・・・y1=x1*W1+b1における[x1の要素数]
* 1つ目の中間層のノード数は？
    * 50個・・・b1の要素数
* 2つ目の中間層のノード数は？
    * 100個・・・b2の要素数
* 出力層のノード数は？
    * 10個・・・b3の要素数

## 予測関数の実装

### [演習]
* 以下のpredict関数を完成させましょう

In [5]:
def sigmoid(x):
    return 1 / (1+ np.exp(-x))

def softmax(x):
    c = np.max(x)    
    exp_a = np.exp(x - c)
    return exp_a / np.sum(exp_a)

def predict(network, x):
    """
    クラス分類するための関数
    network : 重み行列を納めたdict
    x : 入力ベクトル
    return : ソフトマックス関数の結果
    """
    W1, W2, W3 = network["W1"],network["W2"],network["W3"]
    b1, b2, b3 = network["b1"],network["b2"],network["b3"]    
    
    # 順伝播の計算
    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    return 

### 1データずつ予測を行う

In [6]:
# テストデータについて、識別を行う
t = np.argmax(test_labels, axis=1)
p = np.array([]).astype(int)
NUM = 1000
for i in range(NUM):
    y = predict(network, test[i])
    p = np.append(p, np.argmax(y))
    
#　正解データ    
t = np.argmax(test_labels[:NUM], axis=1)

# 予測結果と正解データの比較
print(p)
print(t)

# 識別精度
accuracy_score(t, p)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 

0.085